pkg/commands/credentials.go

```diff
@@ -0,0 +1,299 @@
+// The ask argument will be "username" or "password"
+func (c *OSCommand) DetectUnamePass(command string, ask func(string) string) error {
+	totalListener++
+	currentListener := fmt.Sprintf("%v", totalListener)
+
+	listener := listenerMetaType{AskFunction: ask}
+	listenerMeta[currentListener] = listener
+
+	defer delete(listenerMeta, currentListener)
+
+	end := make(chan error)
+	hostPort := GetFreePort()
+	serverStartedChan := make(chan struct{})
+
+	go runGit(
+		serverStartedChan,
+		end,
+		command,
+		hostPort,
+		currentListener,
+	)
+
+	go runServer(
+		serverStartedChan,
+		end,
+		hostPort,
+		currentListener,
+	)
+
+	err := <-end
+
+	return err
+}
+
+// runGit runs the actual git command with the needed git
+func runGit(serverStartedChan chan struct{}, end chan error, command, hostPort, currentListener string) {
+	<-serverStartedChan
+
+	ex, err := os.Executable()
+	if err != nil {
+		ex = os.Args[0]
+	}
+
+	splitCmd := str.ToArgv(command)
+	cmd := exec.Command(splitCmd[0], splitCmd[1:]...)
+	cmd.Env = append(
+		os.Environ(),
+		"LAZYGIT_HOST_PORT="+hostPort,
+		"LAZYGIT_LISTENER="+currentListener,
+
+		"GIT_ASKPASS="+ex, // tell git where lazygit is located so it can ask lazygit for credentials
+
+		"LANG=en_US.UTF-8",   // Force using EN as language
+		"LC_ALL=en_US.UTF-8", // Force using EN as language
+	)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		outString := string(out)
+		if len(outString) == 0 {
+			end <- err
+			return
+		}
+		end <- errors.New(outString)
+		return
+	}
+	end <- nil
+}
+
+// runServer starts the server that waits for events from the lazygit client
+func runServer(serverStartedChan chan struct{}, end chan error, hostPort, currentListener string) {
+	serverRunning := false
+
+	addy, err := net.ResolveTCPAddr("tcp", "127.0.0.1:"+hostPort)
+	if err != nil {
+		end <- err
+		return
+	}
+
+	inbound, err := net.ListenTCP("tcp", addy)
+	if err != nil {
+		end <- err
+		return
+	}
+
+	go func() {
+		<-end
+		if serverRunning {
+			inbound.Close()
+		}
+	}()
+
+	listener := new(Listener)
+
+	// every listener needs a different name it this is not dune rpc.RegisterName will error
+	err = rpc.RegisterName("Listener"+currentListener, listener)
+	if err != nil {
+		end <- err
+		return
+	}
+
+	serverStartedChan <- struct{}{}
+
+	serverRunning = true
+	rpc.Accept(inbound)
+	serverRunning = false
+}
+
+// GetFreePort returns a free port that can be used by lazygit
+func GetFreePort() string {
+	checkFrom := 5000
+	toReturn := ""
+	for {
+		checkFrom++
+		check := fmt.Sprintf("%v", checkFrom)
+		if IsFreePort(check) {
+			toReturn = check
+			break
+		}
+	}
+	return toReturn
+}
+
+// IsFreePort return true if the port if not in use
+func IsFreePort(port string) bool {
+	conn, err := net.Dial("tcp", "127.0.0.1:"+port)
+	if err == nil {
+		go conn.Close()
+		return false
+	}
+	return true
+}
+
+// SetupClient sets up the client
+// This will be called if lazygit is called through git
+func SetupClient() {
```

Comment:
it would be nice if we could log these errors when they occur rather than fail silently. In the rebasing PR, I put some code in app.go that checks if there is a `LAZYGIT_CLIENT_COMMAND` env variable set, and if so, it does something based on that context. In runGit we could set `LAZYGIT_CLIENT_COMMAND=GET_CREDENTIAL` and then in the Run() method we can go:

```
func (app *App) Run() error {
	if app.ClientContext == "INTERACTIVE_REBASE" {
		return app.Rebase()
	}

	if app.ClientContext == "EXIT_IMMEDIATELY" {
		os.Exit(0)
	}

	if app.ClientContext == "GET_CREDENTIAL" {
		return commands.SetupClient(app.Log)
	}

	return app.Gui.RunWithSubprocesses()
}
```

That way we can log any errors we get to assist in troubleshooting

Reply:
Fixed it :)